encrypted_search/index.py

```diff
@@ -0,0 +1,69 @@
+from typing import List, Set, Tuple
+
+from nltk import word_tokenize
+from nltk.corpus import stopwords
+
+from utils.types import corpus_type, event_type, index_type
+
+
+class EncryptedIndex:
+    """A searchable obfuscated index.
+
+    To prevent data leakage, we obfuscate a simple inverted index into two structures: a collection of arrays and a lookup table. This class sets up this scheme and exposes the data to be searched upon.
+
+    Args:
+        events: List of Matrix room events to be indexed
+
+    Attributes:
+        database: Three dimensional array containing document ids according to obfuscation scheme
+        lookup_table: Map from a keyword to corresponding location in the database
+    """
+
+    def __init__(self, events: List[event_type]):
+        documents, keywords = self.parse(events)
+        inverted_index = self.invert(documents, keywords)
+
+    @staticmethod
+    def parse(events: List[event_type]) -> Tuple[corpus_type, Set[str]]:
+        """Transforms raw Matrix room events into normalized and simplified documents.
+
+        Args:
+            events: List of Matrix room events to be indexed
+
+        Returns:
+            A tuple of the form (D, K), where — D is a mapping from document ids to the set of keywords present in each document and K is a set of all the normalized tokens present in the corpus.
+        """
+
+        documents: corpus_type = {}
+        keywords: Set[str] = set()
+        for event in events:
+            if event["type"] == "m.room.message" \
+                    and "content" in event \
+                    and "body" in event["content"]:
+                event_id: str = event["event_id"]
+                content: str = event["content"]["body"]
+                for punctuation in "!()-[]{};:, <>./?@#$%^&*_~'\"\\":

```

Comment:
This will break the ability to search for Matrix user id mentions, eg "@user:example.com", or email addresses "user@example.com" or even domain names.  Or other things that contain punctuation, like phone numbers or dates, eg 2022-06-16.

I'm not sure what the "right" approach is.  In the past, I've done some hacky things like:
1. Tokenize first
2. Then look at the tokens that you get from the first pass of tokenization.
   1. Strip leading / trailing punctuation -- Although here maybe we want to be wary of removing a leading `@`
   2. There may still be punctuation in the middle of our token, and that's ok.  Maybe it's a Matrix id or an email address, or a domain name etc.  Add it to our ~~list~~ set of keywords.
   3. Then see if we can break up the top-level token into multiple other tokens, by replacing punctuation with spaces and tokenizing again.  So here `@user:example.com` would yield "user", "example", and "com".  And we add all of those tokens to our set of keywords, along with the top-level token "@user:example.com".